language/tools/move-package/src/resolution/resolving_table.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{cmp::Ordering, collections::BTreeMap};
+
+use anyhow::{anyhow, Result};
+use move_core_types::account_address::AccountAddress;
+
+use crate::source_package::parsed_manifest::{NamedAddress, PackageName};
+
+/// A named address qualified with the package name whose scope it belongs to.
+pub type QualifiedAddress = (PackageName, NamedAddress);
+
+/// A data structure for unifying named addresses across packages according to renamings and
+/// assigning them numerical addresses.
+#[derive(Debug)]
+pub struct ResolvingTable {
+    /// Disjoint set data structure for assignments which can either hold no value, a fixed value or
+    /// a forwarding reference to another element in the table.  Each entry in the `redirection`
+    /// table gets a slot in the `assignment` table.
+    assignments: Vec<Assignment>,
+
+    /// Mapping named addresses to an entry in the `assignments` table.
+    redirection: BTreeMap<QualifiedAddress, usize>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum Assignment {
+    Assign(Option<AccountAddress>),
+    Linked(usize),
+}
+
+impl ResolvingTable {
+    /// A fresh `ResolvingTable` with no bindings.
+    pub fn new() -> ResolvingTable {
+        ResolvingTable {
+            assignments: Vec::new(),
+            redirection: BTreeMap::new(),
+        }
+    }
+
+    /// Iterates over the bindings in this table that are within `pkg`'s scope.
+    pub fn bindings(
+        &self,
+        pkg: PackageName,
+    ) -> impl Iterator<Item = (NamedAddress, &Option<AccountAddress>)> {
+        let start = (pkg, NamedAddress::from(""));
+        self.redirection
+            .range(start..)
+            .take_while(move |((scope, _), _)| *scope == pkg)
```

Comment:
Trying to understand this, as I've never used `BtreeMap::range` before but it looks nifty!
Is this more different than `self.redirection.filter(|((scope, _), _)| *scope == pkg)`?

Reply:
It's different from a performance perspective -- range will do a lookup `lg(N)` lookup to find the first element and then continue from there (so if there are `K` entries for a particular package and `N` overall, it will be `O(lg(N) + K)` vs filter's `O(lg(N))`.